### ISSUE TITLE
Improve use of Nu Validator by other projects

### ISSUE THREAD

**Issue by windymilla** [validator/validator#1521]
Project Gutenberg (PG) and Distributed Proofreaders (DP) make extensive use of the Validator in preparation of public domain eBooks. PG uses the Validator as part of the online https://ebookmaker.pglaf.org/ service. DP uses it as part of its final eBook preparation in the offline [Guiguts](https://github.com/DistributedProofreaders/guiguts) tool. Use of the Validator in both organisations has a dual benefit: not only does it ensure the HTML used to generate a range of eBook formats meets currently accepted standards, but it also facilitates the detection of inadvertent errors that could directly affect the accuracy of the displayed text and graphical content. The accuracy of transcription of these public domain books into a free, accessible eBook format is vital for the many users of PG's online library around the world.

The current tagging / release model for the Validator however makes using it very challenging in a few ways:

- The project no longer includes Releases to know which tagged versions are production ready.
- Starting in 2022 even tags are rarely used and the latest tag does not uniquely describe a source of code making it almost impossible to have different consumers sync to a specific version unless they are built from the same git checkout.
- Getting .jar files for tagged versions requires building them yourself which is inconvenient.
- Complicating matters is that the build script always embeds the build time as the version making it impossible to accurately tell what tagged version was actually used to build the binary.
- It's presumably this "built on" date that is output as the version at https://validator.w3.org/nu/ (as of this writing: 23.1.8) and yet there is no tag that defines that version and therefore impossible to have other consumers build that same code.

Here are some ideas on ways to improve this:

- Create and use definitive version numbers that uniquely map to specific code -- probably done with a tag -- this would also be the version embedded in the .jar file and reported on the web page.
    - This would allow consumers to accurately build a version and know what code a binary was built from -- not when whatever code was checked out was compiled.
- Create release builds or tags for production-ready code. Be this through Github releases or through production-specific tags this would allow consumers to know when code was ready for production.

**Comment by eshellman** [comment]
Note that the "most recent release" on github is https://github.com/validator/validator/releases/tag/20.6.30

**Comment by jribbens** [comment]
I'd like to second this, and also mention that publishing the `vnu-jar` npm package more frequently would be good too. It used to be every few months but in recent years has become very infrequent.

**Comment by nivomi** [comment]
I'd also like to say that the release system is a huge boon to staffers working in less github-aware non-tech organizations. Having a specific version to point at for analysis and approvals can be critical, and HTML validation is something that orgs of every scale might need.

**Comment by sideshowbarker** [comment]
> The project no longer includes Releases to know which tagged versions are production ready.
> Create release builds or tags for production-ready code. Be this through Github releases or through production-specific tags this would allow consumers to know when code was ready for production.

The project considers all of its releases to be production-ready.

> Create and use definitive version numbers that uniquely map to specific code … this would also be the version embedded in the .jar file and reported on the web page.

That’s essentially what we we’re now doing.

> Starting in 2022 even tags are rarely used and the latest tag does not uniquely describe a source of code making it almost impossible to have different consumers sync to a specific version unless they are built from the same git checkout.
> It's presumably this "built on" date that is output as the version at [validator.w3.org/nu](https://validator.w3.org/nu/) (as of this writing: 23.1.8) and yet there is no tag that defines that version and therefore impossible to have other consumers build that same code.

The `--version` strings output from the binaries/jars in current releases now always include a commit hash indicating what git HEAD the release was built from. The project’s git repo has a linear history, so the commit hash shown by the `--version` output is functionally the same as a tag. If we were to create tags for each release, they’d simply just be exactly the same as the commit hashes the `--version` output now shows.

> …allow consumers to accurately build a version and know what code a binary was built from -- not when whatever code was checked out was compiled.

Building a particular version from the exact same state of the codebase as shown by the `--version` output from a particular binary/jar can be accomplished by checking-out/switching to the commit hash shown in that `--version` output.

> Starting in 2022 even tags are rarely used and the latest tag does not uniquely describe a source of code making it almost impossible to have different consumers sync to a specific version unless they are built from the same git checkout.

With current releases, different consumers can sync to exactly the same version by examining/comparing the strings output by the `‑‑version` options of the binaries/jars they’re using, and ensuring that those strings exactly match — including the commit-hash part of the string.

**Comment by sideshowbarker** [comment]
@jribbens

> I'd like to second this, and also mention that publishing the `vnu-jar` npm package more frequently would be good too. It used to be every few months but in recent years has become very infrequent.

We now release a new npm package every time we push code to GitHub. You can find latest npm package here:

https://github.com/validator/validator/pkgs/npm/vnu-jar

**Comment by jribbens** [comment]
@sideshowbarker 
> The project considers all of its releases to be production-ready.

Until 12 hours ago you had no releases in the last 5 years ;-)

I did wonder if your policy was "if it's been pushed to main then it's production-ready" but without an explicit statement to that effect, it's not a safe thing to assume.

> We now release a new npm package every time we push code to GitHub.

There's something stuck in your process then - the latest version on npmjs.com, which gets 13,000 downloads a week, is over a year old. (For some reason, even older versions get far more downloads.)

**Comment by sideshowbarker** [comment]
> [@sideshowbarker](https://github.com/sideshowbarker)
> 
> > The project considers all of its releases to be production-ready.
> 
> Until 12 hours ago you had no releases in the last 5 years ;-)

Point taken. I’d meant the releases we’d still been doing at https://github.com/validator/validator/releases

For updated details on what the project’s now doing for Maven releases:

https://github.com/validator/validator/issues/1354#issuecomment-3471627259

> I did wonder if your policy was "if it's been pushed to main then it's production-ready" but without an explicit statement to that effect, it's not a safe thing to assume.

Yeah, "if it's been pushed to main then it's production-ready" is now the policy

> > We now release a new npm package every time we push code to GitHub.
> 
> There's something stuck in your process then - the latest version on npmjs.com, which gets 13,000 downloads a week, is over a year old. (For some reason, even older versions get far more downloads.)

I’m not doing releases to npmjs.com any longer — at least not until I can manage to figure out how to fully automate the releases using GitHub Actions.

But we do release an updated npm package continuously at https://github.com/validator/validator/pkgs/npm/vnu-jar

That said, see also https://github.com/validator/validator/issues/1470 — which takes a position (which I don’t disagree with) that:

1. _“one would expect that installing a dependency through a dependency manager automatically resolves all dependencies”_, and that our npm package
1.  But our npm package — in its current form — doesn’t meet that expectation, because it currently depends on the user already having a JVM installed, and doesn’t automatically install a JVM for the user.

…and probably most users anyway don’t want some npm package causing an entire JVM — or really, any other non-Node/JavaScript runtime — to be installed on their systems.

In other words, the npm package in its current form is arguably of kind of marginal value.

What it better should do instead is, to:

1. Figure out the system architecture (Linux, macOS, or Windows) of the user’s system.
2. Install the correct binary runtime image from https://github.com/validator/validator/releases/tag/latest that is compiled for the user’s architecture.

But that’s all a separate thing that what this issue here is about, I guess.

**Comment by sideshowbarker** [comment]
Hopefully my comments at https://github.com/validator/validator/issues/1521#issuecomment-3448024995 and elsewhere have clarified things. Regardless, for now I can’t see any further action that I’m planning to take beyond what I’ve described already — so I’m going ahead and closing this issue.

But anyone should feel free to continue to comment here — and we could re-open this issue if it ends up seeming like we should.

**Comment by jribbens** [comment]
> Hopefully my comments at [#1521 (comment)](https://github.com/validator/validator/issues/1521#issuecomment-3448024995) and elsewhere have clarified things. Regardless, for now I can’t see any further action that I’m planning to take beyond what I’ve described already — so I’m going ahead and closing this issue.
> 
> But anyone should feel free to continue to comment here — and we could re-open this issue if it ends up seeming like we should.

I would just say that the install instructions at https://github.com/validator/validator/pkgs/npm/vnu-jar (which say to do `npm install @validator/vnu-jar@25.10.31`) don't work, so that probably needs changing (in fact I can't find any way to download anything from that page or anywhere linked from it), and if you're not going to release any more `vnu-jar` npm packages then perhaps you should update the README on that package to indicate it is deprecated and to provide instructions on the correct place to look.

**Comment by windymilla** [comment]
Thanks for the clarifications, etc., @sideshowbarker 

In a rewrite of one of our book creation/checking tools, we have switched to using the validator via the API, which works well for us and avoids the above issues, since then we know we are always using the production version.

**Comment by sideshowbarker** [comment]
@jribbens I have auto-publishing to npmjs.com working now. So please try it — and if you run into any problems, raise new/separate issues.

One problem I’m aware of is that at https://www.npmjs.com/package/vnu-jar and in the package itself, it’s not using the right readme content. It should be using https://github.com/validator/validator/blob/main/npm.md for that. But that’s relatively easy to fix, and I should be able to get it fixed by tomorrow or so.

**Comment by jribbens** [comment]
@sideshowbarker Seems good so far, just installed `vnu-jar@25.11.4` and my tests passed...
